HerPortal.BusinessLogic/Services/CsvFileService/CsvFileService.cs

```diff
@@ -1,75 +1,83 @@
-        // Notably, we can't confirm a download, so it's possible that we mark a file as downloaded
-        //   but the user has some sort of issue and doesn't get it
-        // We put this line as late as possible in the method for this reason
-        await dataAccessProvider.MarkCsvFileAsDownloadedAsync(custodianCode, year, month, userId);
-
-        return fileStream;
-    }
-}
+using System.Security;
+using HerPortal.BusinessLogic.ExternalServices.S3FileReader;
+using HerPortal.BusinessLogic.Models;
+using HerPortal.Data;
+using HerPublicWebsite.BusinessLogic.Services.S3ReferralFileKeyGenerator;
+
+namespace HerPortal.BusinessLogic.Services.CsvFileService;
+
+public class CsvFileService : ICsvFileService
+{
+    private readonly IDataAccessProvider dataAccessProvider;
+    private readonly S3ReferralFileKeyService keyService;
+    private readonly IS3FileReader s3FileReader;
+
+    public CsvFileService
+    (
+        IDataAccessProvider dataAccessProvider,
+        S3ReferralFileKeyService keyService,
+        IS3FileReader s3FileReader
+    ) {
+        this.dataAccessProvider = dataAccessProvider;
+        this.keyService = keyService;
+        this.s3FileReader = s3FileReader;
+    }
+    
+    public async Task<IEnumerable<CsvFileData>> GetByCustodianCodesAsync(IEnumerable<string> custodianCodes, int userId)
+    {
+        var downloads = await dataAccessProvider.GetLastCsvFileDownloadsAsync(userId);
+        var files = new List<CsvFileData>();
+
+        foreach (var custodianCode in custodianCodes)
+        {
+            var s3Objects = await s3FileReader.GetS3ObjectsByCustodianCodeAsync(custodianCode);
+            files.AddRange(s3Objects.Select(s3O =>
+                {
+                    var data = keyService.GetDataFromS3Key(s3O.Key);
+                    var downloadData = downloads.SingleOrDefault(d =>
+                        d.CustodianCode == data.CustodianCode
+                        && d.Year == data.Year
+                        && d.Month == data.Month
+                    );
+                    return new CsvFileData
+                    (
+                        data.CustodianCode,
+                        data.Month,
+                        data.Year,
+                        s3O.LastModified,
+                        downloadData?.LastDownloaded
+                    );
+                }
+            ));
+        }
+
+        return files
+            .OrderByDescending(f => new DateOnly(f.Year, f.Month, 1))
+            .ThenBy(f => LocalAuthorityData.LocalAuthorityNamesByCustodianCode[f.CustodianCode]);
+    }
+    
+    public async Task<Stream> GetFileForDownloadAsync(string custodianCode, int year, int month, string userEmailAddress)
+    {
+        // Important! First ensure the logged-in user is allowed to access this data
+        var userData = await dataAccessProvider.GetUserByEmailAsync(userEmailAddress);
+        if (userData.LocalAuthorities.All(la => la.CustodianCode != custodianCode))
```

Comment:
(FWIW I think this was more natural with the negation pulled out, but it's not too important - I'd say 'Not any of your custodian codes match this one' rather than 'All your custodian codes do not match this one')